### PR TITLE
[Maintenance] Allow to choose whether use Webpack or Gulp

### PR DESCRIPTION
**Pull request by jakubtobiasz**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.12                   |
| Bug fix?        | no                                                       |
| New feature?    | yes                                                       |
| BC breaks?      | kinda, FE is not under BC policy but technically it can break the app just after the upgrade            |
| Deprecations?   | kinda, gulp is deprecated from now  |
| Related tickets | #13900                     |
| License         | MIT                                                          |

tl;dr
I've switched to webpack in every possible place. Also, I've replaced the old webpack test build with the gulp test build. Docs update is coming in the next PR.

### CODE REVIEW

File: .github/workflows/application.yml [hunk 1]
```diff
@@ -242,6 +242,7 @@ jobs:
                 php: ["8.0"]
                 symfony: ["^5.4"]
                 postgres: ["13.3"]
+                node: ["16.x"]
```
Comment (jakubtobiasz) [inline .github/workflows/application.yml:245]: fyi I've used `16.x` as it is LTS version and previously we didn't build assets in non-JS tests but now we have to. `16.x` and `18.x` are run in JS tests.

Comment (coldic3) [reply .github/workflows/application.yml:245]: What will be the difference between this one and `test-application-with-frontend`?

File: .gitignore [hunk 2]
```diff
@@ -51,3 +51,10 @@
 /config/jwt/*.pem
 !/config/jwt/*-test.pem
 ###< lexik/jwt-authentication-bundle ###
+
+###> symfony/webpack-encore-bundle ###
+/node_modules/
```
Comment (coldic3) [inline .gitignore:55]: `node_modules` dir is already ignored in line 15.

File: .github/workflows/application.yml (hunk 1 above)
Comment (jakubtobiasz) [reply .github/workflows/application.yml:245]: @coldic3 `with-frontend` has a misleading name as it runs tests having a `@javascript` tag. Here we need assets to satisfy configuration. So the difference is significant as we run different sets of tests.

File: UPGRADE-1.12.md [hunk 3]
```diff
@@ -115,3 +111,29 @@ Therefore you need to update your code to follow this change.
...
+
+Example changes we made in our codebase to adjust it to the new version of dependencies might be found [here](https://github.com/Sylius/Sylius/pull/14319/files).
+Remember, when you face any issues while updating your dependencies you might ask for help in our [Slack](https://sylius-devs.slack.com/) channel.
+
+#### Webpack becomes our default build tool 🎉
+
+`1.12` comes with a long-awaited change - Webpack becomes our default build tool. 
+
+If you want to stay with Gulp, you can do it by following the steps below:
+1. Go to `config/packages/_sylius.yaml` file and add the following line:
```
Comment (GSadee) [inline UPGRADE-1.12.md:134]:
```suggestion

1. Go to `config/packages/_sylius.yaml` file and add the following line:
```

File: src/Sylius/Bundle/UiBundle/Resources/config/app/parameters.yaml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# This file is part of the Sylius package.
+# (c) Paweł Jędrzejewski
+
+parameters:
+    sylius_ui.use_webpack: true
```
Comment (GSadee) [inline src/Sylius/Bundle/UiBundle/Resources/config/app/parameters.yaml]: I would be for defining this parameter in the configuration tree in UiBundle and then change:
```suggestion
sylius_ui:
    use_webpack: true
```

Comment (jakubtobiasz) [reply src/Sylius/Bundle/UiBundle/Resources/config/app/parameters.yaml]: @GSadee I wanted to do it this way but I also declare
```yaml
twig:
    globals:
        use_webpack: '%sylius_ui.use_webpack%'
```

When I was setting the parameter in the extension it was complaining about a missing parameter. I've decided if I need to define it anyway to drop the idea about configuration and left only the parameter.

File: src/Sylius/Bundle/ShopBundle/Resources/views/Layout/Header/_logo.html.twig [hunk 5]
```diff
@@ -1,3 +1,11 @@
...
-</div>
+{% if use_webpack %}
+    <div class="column">
+        <a href="{{ path('sylius_shop_homepage') }}">
+            <img src="{{ asset('build/shop/images/logo.png', 'shop') }}" alt="Sylius logo" class="ui small image" />
+        </a>
+    </div>
+{% else %}
+    <div class="column">
+        <a href="{{ path('sylius_shop_homepage') }}"><img src="{{ asset('assets/shop/img/logo.png') }}" alt="Sylius logo" class="ui small image" /></a>
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Layout/Header/_logo.html.twig]:
```suggestion
        <a href="{{ path('sylius_shop_homepage') }}">
            <img src="{{ asset('assets/shop/img/logo.png') }}" alt="Sylius logo" class="ui small image" />
        </a>
```

File: src/Sylius/Bundle/AdminBundle/Resources/views/Layout/_logo.html.twig [hunk 6]
```diff
@@ -1,5 +1,13 @@
-<a class="item" href="{{ path('sylius_admin_dashboard') }}" style="padding: 13px 0;">
-    <div style="max-width: 90px; margin: 0 auto;">
-        <img src="{{ asset('assets/admin/img/admin-logo.svg') }}" class="ui fluid image">
-    </div>
-</a>
+{% if use_webpack %}
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Layout/_logo.html.twig]: Just wanted to mention that you could use `set` with ternary for those, but on second thought `if`/`else` may be more readable.

File: .github/workflows/application.yml [hunk 7]
```diff
@@ -299,6 +300,40 @@ jobs:
...
+            -
+                name: Get Yarn cache directory
+                id: yarn-cache
+                run: echo "::set-output name=dir::$(yarn cache dir)"
+
+            -
+                name: Cache Yarn
+                uses: actions/cache@v2
+                with:
+                    path: ${{ steps.yarn-cache.outputs.dir }}
```
Comment (Ferror) [inline .github/workflows/application.yml:318]: Do we have to specify it for a "without-frontend" actions?

File: .github/workflows/application.yml [hunk 8]
```diff
@@ -500,6 +570,7 @@ jobs:
                 mysql: ["5.7", "8.0"]
                 twig: ["3.x"]
                 dbal: ["3"]
+                node: ["16.x"]
```
Comment (Ferror) [inline .github/workflows/application.yml:572]: W should also test with Node 14. I remember that I have somewhere a PR with Node 16 tests in Sylius Plus and they are failing bit time :(

BTW. Current Sylius-Standard is Node 14 driven too

Comment (Ferror) [reply .github/workflows/application.yml:572]: Unless we want to upgrade it, which is cool. But we have to state it :)

File: src/Sylius/Bundle/AdminBundle/Resources/views/Layout/_logo.html.twig (hunk 6 above)
Comment (jakubtobiasz) [reply src/Sylius/Bundle/AdminBundle/Resources/views/Layout/_logo.html.twig]: I prefer to omit `set` whenever it is possible (I mean whenever it doesn't improve readability significantly).

File: .github/workflows/application.yml (hunk 8 above)
Comment (jakubtobiasz) [reply .github/workflows/application.yml:572]: @Ferror as we're bumping all deps I've decided to stay with LTS and the current version for node. In this case, we use it to build assets as we need them (every time we touch the container we need to have them as we have configured assets pointing to `manifest.json` files).

File: .github/workflows/application.yml (hunk 1 above)
Comment (coldic3) [reply .github/workflows/application.yml:245]: If the only difference is with running different sets of tests maybe we can run them all in the same pipeline 🤔

File: src/Sylius/Bundle/UiBundle/Resources/config/app/parameters.yaml (hunk 4 above)
Comment (GSadee) [reply src/Sylius/Bundle/UiBundle/Resources/config/app/parameters.yaml]: It would probably work if you set this parameter in `prepend` method in the extension 🤔

File: Makefile [hunk 9]
```diff
@@ -29,7 +29,7 @@ backend:
 
 frontend:
 	yarn install --pure-lockfile
-	GULP_ENV=prod yarn build
+	yarn encore production
```
Comment (Zales0123) [inline Makefile:32]: Btw, are we using this Makefile anywhere? Maybe we should include it in our pipeline, so it would be tested as well? 🤔

File: UPGRADE-1.12.md [hunk 10]
```diff
@@ -115,3 +111,29 @@ Therefore you need to update your code to follow this change.
    + "rollup-plugin-uglify": "^6.0.2",
    ```
     Please visit [3.x Migration Guide](https://www.chartjs.org/docs/latest/getting-started/v3-migration.html) for more information.
+
+Example changes we made in our codebase to adjust it to the new version of dependencies might be found [here](https://github.com/Sylius/Sylius/pull/14319/files).
+Remember, when you face any issues while updating your dependencies you might ask for help in our [Slack](https://sylius-devs.slack.com/) channel.
+
+#### Webpack becomes our default build tool 🎉
+
+`1.12` comes with a long-awaited change - Webpack becomes our default build tool.
```
Comment (Zales0123) [inline UPGRADE-1.12.md:130]: Even though I share the enthusiasm 😄 I would keep upgrade files as _rough_ as possible (purely to make it only informative) and keep the emotions for blogposts or tweets 🚀

File: src/Sylius/Bundle/UiBundle/Resources/config/app/parameters.yaml (hunk 4 above)
Comment (Zales0123) [reply src/Sylius/Bundle/UiBundle/Resources/config/app/parameters.yaml]: I would be 👍 for configuration, as this is indeed what we're doing here - **configuring** how the UI of Sylius should work. We can process this configuration in the bundle extension and make a parameter from it (we're already doing such things for the `sylius_api.enabled` parameter), so it would be consistent with our current implementation and valid from the functional point of view 🖖

**Review by Zales0123 — COMMENTED** [review]
Finally we're doing something with it 💪

File: Makefile (hunk 9 above)
Comment (jakubtobiasz) [reply Makefile:32]: @Zales0123 it is used in Zbyszek's Docker.

File: UPGRADE-1.12.md (hunk 10 above)
Comment (jakubtobiasz) [reply UPGRADE-1.12.md:130]: GitHub doesn't show this is outdated but I've changed it 💃🏻

File: .github/workflows/application.yml (hunk 1 above)
Comment (jakubtobiasz) [reply .github/workflows/application.yml:245]: @coldic3 our Behats are already unstable, I prefer to keep them split 😂

**Review by GSadee — APPROVED** [review]

File: .github/workflows/application.yml (hunk 8 above)
Comment (lchrusciel) [reply .github/workflows/application.yml:572]: but some checks of the correctness of this change with plus and sylius standard should be done. Also, to check if it works there and how much hustle we have to change it

File: config/packages/webpack_encore.yaml [hunk 11]
```diff
@@ -0,0 +1,5 @@
+webpack_encore:
```
Comment (lchrusciel) [inline config/packages/webpack_encore.yaml]: Should it be part of a distribution or do we want everyone to define the defaults, that you've put here? With this config, Sylius (in this repo) will work, but everyone using it in vendor won't access this file (like Sylius-Standard)

File: src/Sylius/Bundle/AdminBundle/Resources/views/Layout/_logo.html.twig [hunk 12]
```diff
@@ -1,5 +1,13 @@
...
+    <a class="item" href="{{ path('sylius_admin_dashboard') }}" style="padding: 13px 0;">
+        <div style="max-width: 90px; margin: 0 auto;">
+            <img src="{{ asset('build/admin/images/admin-logo.svg', 'admin') }}" class="ui fluid image">
+        </div>
+    </a>
+{% else %}
+    <a class="item" href="{{ path('sylius_admin_dashboard') }}" style="padding: 13px 0;">
+        <div style="max-width: 90px; margin: 0 auto;">
+            <img src="{{ asset('assets/admin/img/admin-logo.svg') }}" class="ui fluid image">
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Layout/_logo.html.twig:6]: Isn't path change a BC Break? Can't we maintain an old path to simplify migration? At the end, it is a smell for me, if we have to have if statements in templates to make it work

File: .github/workflows/application.yml (hunk 8 above)
Comment (jakubtobiasz) [reply .github/workflows/application.yml:572]: I've left Node 16 for **Non JS** tests as they're used only to satisfy webpack encore configuration (manifest.json must be present). For **JS** configuration I've added node 14.x.

File: src/Sylius/Bundle/AdminBundle/Resources/views/Layout/_logo.html.twig (hunk 12 above)
Comment (jakubtobiasz) [reply src/Sylius/Bundle/AdminBundle/Resources/views/Layout/_logo.html.twig:6]: I think we discussed it on the morning meeting. The paths are different and most of community using webpack is already using the new path so there's possibility we would introduce more problems changing the path for webpack.

File: config/packages/webpack_encore.yaml (hunk 11 above)
Comment (jakubtobiasz) [reply config/packages/webpack_encore.yaml]: Distributed between bundles

File: UPGRADE-1.12.md [hunk 13]
```diff
@@ -84,24 +84,20 @@ with test or remove these services with complier pass.
...
 
-### Asset management changes
+### Frontend toolset changes
+
+#### Dependencies update
+
+In `1.12` we have updated our frontend dependencies to the latest versions. This means that you might need to update your dependencies as well.
+The full list of all dependencies can be found in the [package.json](./package.json) file.
+
+Due to a fact every project is different, we cannot provide a list of all changes that might be needed. However, we have prepared a list of the most common changes that might be needed.
```
Comment (NoResponseMate) [inline UPGRADE-1.12.md]:
```suggestion
Because every project is different, we cannot provide a list of all changes that might be needed. However, we have prepared a short list of fixes for the most common issues.
```

File: src/Sylius/Bundle/AdminBundle/Resources/views/Security/_content.html.twig [hunk 14]
```diff
@@ -1 +1,15 @@
-{% include '@SyliusUi/Security/_login.html.twig' with {'action': path('sylius_admin_login_check'), 'paths': {'logo': 'assets/admin/img/logo.png'}} %}
+{% if use_webpack %}
+    {% include '@SyliusUi/Security/_login.html.twig'
+        with {
+        'action': path('sylius_admin_login_check'),
+        'paths': {'logo': asset('build/admin/images/logo.png', 'admin')}
+    }
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Security/_content.html.twig]:
```suggestion
        with {
            'action': path('sylius_admin_login_check'),
            'paths': {'logo': asset('build/admin/images/logo.png', 'admin')}
        }
```

**Review by NoResponseMate — APPROVED** [review]

File: UPGRADE-1.12.md [hunk 15]
```diff
@@ -125,3 +121,47 @@ Therefore you need to update your code to follow this change.
...
+
+1. Go to `config/packages/_sylius.yaml` file and add the following line:
+   ```yaml
+   sylius_ui:
+      use_webpack: false
+   ```
+
+If you decide to use Webpack, you need to follow the steps below:
+
+1. Make sure you have latest js dependencies installed (you can compare your `package.json` file with the one from `1.12`).
```
Comment (GSadee) [inline UPGRADE-1.12.md:142]:
```suggestion
1. Make sure you have the latest js dependencies installed (you can compare your `package.json` file with the one from `1.12`).
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, Jakub! :tada:
